### PR TITLE
Add limits to OTEL attribute count and length

### LINKED ISSUE #2066: Configuration for the dashboard to enforce OTEL attribute limits
The dashboard stores telemetry in bounded collections. However, an individual piece of telemetry could still be large, e.g. 1000 attributes with keys are values 100 megabytes each.

SDKs are encouraged to provide [attribute limits](https://opentelemetry.io/docs/specs/otel/common/#attribute-limits). The dashboard should have the option to enforce limits on incoming telemetry to avoid exessive memory use.

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/aspire/issues/2066

* Apply limits to attributes, re: https://opentelemetry.io/docs/specs/otel/common/#attribute-limits
* Apply limits to spans, re: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#span-limits
* Configuration values to customize attribute limits
* Reduced maximum metric count and max duration is now 12 hours
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2531)

### CODE REVIEW

File: src/Aspire.Dashboard/Otlp/Storage/TelemetryRepository.cs [hunk 1]
```diff
@@ -14,25 +14,40 @@
...
+{
+    internal const string LogCountLimitKey = "DOTNET_DASHBOARD_OTEL_LOG_COUNT_LIMIT";
+    internal const string TraceCountLimitKey = "DOTNET_DASHBOARD_OTEL_TRACE_COUNT_LIMIT";
+    internal const string MetricsCountLimitKey = "DOTNET_DASHBOARD_OTEL_METRIC_COUNT_LIMIT";
+    internal const string AttributeCountLimitKey = "DOTNET_DASHBOARD_OTEL_ATTRIBUTE_COUNT_LIMIT";
+    internal const string AttributeLengthLimitKey = "DOTNET_DASHBOARD_OTEL_ATTRIBUTE_LENGTH_LIMIT";
+    internal const string SpanEventCountLimitKey = "DOTNET_DASHBOARD_OTEL_SPAN_EVENT_COUNT_LIMIT";
+
+    private const int DefaultTraceCountLimit = 10_000;
+    private const int DefaultMetricsCountLimit = 50_000; // Allows for 1 metric point per second for over 12 hours.
```
Comment (danmoseley) [inline src/Aspire.Dashboard/Otlp/Storage/TelemetryRepository.cs:37]: Do we know the dashboard is still using a reasonable amount of memory when these limits are both reached (max items all max length)?

Comment (JamesNK) [reply src/Aspire.Dashboard/Otlp/Storage/TelemetryRepository.cs:37]: No. I'm sure that someone who really wants to max memory could use the worst-case scenario (10,000 log messages, each with 128 attributes, each with a large value) to use a lot of memory.

However, the limits are all configurable, so a hoster has the option to place smaller limits.

The limits will help prevent one errant log message (megabyte sized message) from using excessive memory and keep the dashboard usable.

**Comment by mitchdenny** [comment]
I'm not sure if we already have this somewhere else (I assumed it would be in this PR) but how do we tell the user that we've dropped their message because it is too large. Perhaps we should give them a heads up that we've dropped a message and perhaps the first part of the message to help them track down where it is coming from in their own code?

**Comment by JamesNK** [comment]
It's normal for telemetry stores to have maximum lengths and automatically trim data. We could make it so there is a log message at a low level (Debug?) when trimming happens. You would need to update logging to see it.

Do you think that's necessary/valuable? IMO we can wait for feedback that it's confusing before doing anything.

**Comment by JamesNK** [comment]
Review please 🙏

**Comment by JamesNK** [comment]
> I see this truncates long strings/excess attributes. When there are simply too many events do we simply drop the message, or drop from front to make room?

There is already code to limit the total number of logs/traces/metrics. When the limit is reached (i.e. 10,000 logs) then the oldest is dropped to make room.

**Review by tlmii — APPROVED** [review]
The OTEL stuff is a little out of my wheelhouse, but what you're doing makes sense and nothing jumps out at me. Will be good to have this protection in place.

**Pull request merged**